### ISSUE TITLE
Dual Gpu card single platform different performance for each GPU

### ISSUE THREAD

**Issue by paolodalberto** [CNugteren/CLBlast#364]
This is an example for device 1 (using tuning)
[clblast_xgemm_2_32.json.txt](https://github.com/CNugteren/CLBlast/files/3399840/clblast_xgemm_2_32.json.txt)
[sgemm_plot.pdf](https://github.com/CNugteren/CLBlast/files/3399844/sgemm_plot.pdf)

this is for device 0 (using tuning)
[clblast_xgemm_1_32.d0.json.txt](https://github.com/CNugteren/CLBlast/files/3399841/clblast_xgemm_1_32.d0.json.txt)
[hgemm_plot.pdf](https://github.com/CNugteren/CLBlast/files/3399845/hgemm_plot.pdf)

**Comment by paolodalberto** [comment]
insert the debug for the two devices (default)

[out.1.txt](https://github.com/CNugteren/CLBlast/files/3400372/out.1.txt)
[out.txt](https://github.com/CNugteren/CLBlast/files/3400374/out.txt)

**Comment by paolodalberto** [comment]
let me know if you need any further information. 
thanks

**Comment by paolodalberto** [comment]
the same identical codes are compiled and with no options ...
[out.txt](https://github.com/CNugteren/CLBlast/files/3409688/out.txt)
[out.1.txt](https://github.com/CNugteren/CLBlast/files/3409689/out.1.txt)

**Comment by paolodalberto** [comment]
Same codes are compiled with the same options (of course different execution time)

[out.txt](https://github.com/CNugteren/CLBlast/files/3409696/out.txt)
[out.1.txt](https://github.com/CNugteren/CLBlast/files/3409697/out.1.txt)

**Comment by CNugteren** [comment]
Sorry I'm not at home at the moment, I'll take a look soon when i'm back.

**Comment by paolodalberto** [comment]
no problemo. I do not think it is a problem of CLBlas but of the opencl.
I am working on a priority queue, that will give less priority for slower GPUs just in case this run time issues will continue in the future.   It was due to use better the resources. 

Enjoy your time (vacation ?) and let me know your ideas when possible.

**Comment by CNugteren** [comment]
I doubt indeed that this is a CLBlast issue. A couple of remarks regarding your question:

1) If you have two devices with the same name, you should just run the tuner once, only one of the results will be stored and used anyway: it is name based.
2) From your tuning results I see that both results get almost the same parameters as best (so probably both are roughly equal) and that both best times are 0.6 ms, so that matches as well. In other words, both your devices seem to behave the same.
3) A 0.6 ms result for 1024*1024*1024 in the tuner should give you ~3500 GFLOPS at best, but with some overhead in the library and some other pre/post kernels your results in the second graph of around 3000 GFLOPS seem to match and be correct. So the issue is with the other device (device 1).
4) Your other device reaches only ~1000 GFLOPS, so something is wrong there. But it did well when running the tuner, so it is perhaps a driver/heat/clock thingy? Don't think this is CLBlast related.
5) In general regarding the sudden jump in your speed graphs you can probably benefit from running the xgemm_direct tuner and then subsequently (after updating the DB) also the routine_xgemm kernel. See more info about the two methods [here](https://github.com/CNugteren/CLBlast/blob/master/doc/details_gemm.md).

**Comment by paolodalberto** [comment]
The last test I did, I checked if at run time the two GPUs get the same code (binary kernel) and they do. As they should because the two GPUs are in the same card and they are the same.  Thus the performance difference is not related by CLBlast.  This is interesting and unpleasant. 

For me, this means I cannot take for granted the performance and I need to create a priority Queue among the GPUs so that the fastest will work more. More work but it is good work.  I implemented already and I will share the results. For the moment I will close the issue.  

thanks
